server/service/src/sync/translations/special/name_merge.rs

```diff
@@ -56,7 +60,89 @@ impl SyncTranslation for NameMergeTranslation {
+        // The remaining NSJ that we keep must logically OR each of these fields with the corresponding field in the deleted NSJs.
+        // We prefer making the name visible to stores rather than losing visibility as it allows users to still make invoices and orders
+        let store_repo = StoreRepository::new(connection);
+        let store = store_repo
+            .query_one(StoreFilter::new().name_id(EqualFilter::equal_to(&data.merge_id_to_keep)))?;
+        let deletes = name_store_joins_for_delete
+            .iter()
+            .filter_map(|nsj_delete| {
+                // delete nsj_delete if it points to the store that belongs to the "keep" name. Avoids:
+                // storeK.name_id == nameK.id
+                // storeK joined to nameD
+                // nameD merged into nameK
+                // storeK joined to nameK (delete the join before this happens, stores shouldn't be visible to themselves)
+                if let Some(store) = &store {
+                    if nsj_delete.name_store_join.store_id == store.store_row.id {
+                        return Some(PullDeleteRecord {
+                            id: nsj_delete.name_store_join.id.clone(),
+                            table: PullDeleteRecordTable::NameStoreJoin,
+                        });
+                    }
+                }
+
+                // Delete duplicate name_store_joins. Avoids:
+                // ("joined to" meaning store->nsj->name_link->name)
+                // storeA joined to nameK
+                // storeA joined to nameD
+                // storeB joined to nameD
+                // nameD merged into nameK
+                // storeA joined to nameK
+                // storeA joined to nameK (delete this join to avoid showing twice in lists seemingly as a duplicate)
+                // storeB joined to nameK (make sure we don't accidentally delete this one, or visibility of nameK will be lost for storeB)
+                if let Some(nsj_keep) = name_store_joins_for_keep.iter().find(|nsj_keep| {
+                    nsj_keep.name_store_join.store_id == nsj_delete.name_store_join.store_id
+                }) {
+                    // We must also consider nsj_delete.name_is_customer and nsj_delete.name_is_supplier.
+                    // The remaining NSJ that we keep must logically OR each of these fields with the corresponding field in the deleted NSJs.
+                    // We prefer making the name visible to stores rather than losing visibility as it allows users to still make invoices and orders
+                    if (!nsj_keep.name_store_join.name_is_customer
+                        && nsj_keep.name_store_join.name_is_customer)
+                        || (!nsj_keep.name_store_join.name_is_supplier
+                            && nsj_keep.name_store_join.name_is_supplier)
+                    {
+                        upserts.push(PullUpsertRecord::NameStoreJoin(NameStoreJoinRow {
```

Comment:
@clemens-msupply I personally feel I have done something illegal by having a `Vec.push()` side effect in the map function lol.

Reply:
Oh yeah good point :) Alternative is to just fall back to a simple for loop. Main thing for me was to only iterate `name_store_joins_for_delete` once to make it clear items are not deleted twice...

Reply:
Yea I think that is a solid win from the filter_map. I'm happy with it as it is now though if you are.